packages/codemirror/src/mode.ts

```diff
@@ -3,199 +3,1603 @@
-import 'codemirror/mode/sql/sql';
-import './codemirror-ipython';
-import './codemirror-ipythongfm';
-
-// Stub for the require function.
-declare let require: any;
-
-/**
- * The namespace for CodeMirror Mode functionality.
- */
+import {
+  defaultHighlightStyle,
+  LanguageDescription,
+  LanguageSupport,
+  StreamLanguage,
+  StreamParser
+} from '@codemirror/language';
+
+import { highlightTree } from '@lezer/highlight';
+
+// This ensures the language spec for python will be loaded when
+// we instantiate a new editor instance, which is required since
+// python is the default language and we don't want to split
+// the editor constructor because of asynchronous loading.
+import { python } from '@codemirror/lang-python';
+
 export namespace Mode {
   /**
-   * The interface of a codemirror modeInfo spec.
+   * The interface of a codemirror language spec.
    */
   export interface ISpec {
-    ext?: string[];
-    name?: string;
-    mode: string;
-    mime: string;
+    name: string;
+    alias?: readonly string[];
+    mime: string | readonly string[];
+    load?: () => Promise<LanguageSupport>;
+    extensions?: readonly string[];
+    filename?: RegExp;
+    support?: LanguageSupport;
   }
 
-  /**
-   * The interface of a codemirror mode spec.
-   */
-  export interface IMode {
-    name: string;
-    [key: string]: JSONValue;
+  export function legacy(parser: StreamParser<unknown>): LanguageSupport {
+    return new LanguageSupport(StreamLanguage.define(parser));
   }
 
-  /**
-   * The interface for a codemirror spec resolver.
-   */
-  export interface ISpecLoader {
-    /**
-     * A function which returns whether it was successfully loaded
-     */
-    (spec: ISpec): Promise<boolean>;
+  function sql(dialectName: keyof typeof import('@codemirror/lang-sql')) {
+    return import('@codemirror/lang-sql').then(m =>
+      m.sql({ dialect: (m as any)[dialectName] })
+    );
   }
 
-  const specLoaders: Private.IRankItem[] = [
-    {
-      // Simplest, cheapest check by mode name.
-      loader: async spec => CodeMirror.modes.hasOwnProperty(spec.mode),
-      rank: 0
-    },
-    {
-      // Fetch the mode asynchronously.
-      loader: function (spec) {
-        return new Promise<boolean>((resolve, reject) => {
-          // An arrow function below seems to miscompile in our current webpack to
-          // invalid js.
-          require([
-            `codemirror/mode/${spec.mode}/${spec.mode}.js`
-          ], function () {
-            resolve(true);
-          });
-        });
-      },
-      rank: 99
-    }
+  // Code mirror uses two similar structures, a plain object with optional fields,
+  // and a class with the same fields but all mandatory. Maybe adopting the same
+  // pattern would be less confusing (although far more verbose)
+  function makeSpec(spec: ISpec): ISpec {
+    let res = LanguageDescription.of(spec) as unknown as ISpec;
+    res.mime = spec.mime;
+    return res;
+  }
+
+  const goldorak = [
```

Comment:
Let's use a more appropriate name :wink:

Reply:
Indeed this was "review quality indicator" :D